plaidml/edsl/edsl_test.py

```diff
@@ -258,6 +258,31 @@ class TestEdsl(unittest.TestCase):
     def compare_results(self, program, expected):
         self.assertMultiLineEqual(str(program).strip(), expected.strip())
 
+    def test_lower_precision(self):
+        I = Tensor(LogicalShape(plaidml.DType.FLOAT64, [3, 3]))
+        O = I + 1 + 2.0
+
+        program = Program(
+            'lower_precision', [O], floatx=plaidml.DType.FLOAT32, intx=plaidml.DType.INT32)
+        expected = '''
+
+!i32 = type tensor<!eltwise.i32>
+!f32 = type tensor<!eltwise.f32>
+module {
+  func @lower_precision(%arg0: tensor<3x3x!eltwise.f64>) -> tensor<3x3x!eltwise.f64> {
+    %c1 = "eltwise.sconst"() {value = 1 : i64} : () -> !i32
+    %cst = "eltwise.sconst"() {value = 2.000000e+00 : f64} : () -> !f32
+    %0 = "eltwise.add"(%arg0, %c1) : (tensor<3x3x!eltwise.f64>, !i32) -> tensor<3x3x!eltwise.f64>
+    %1 = "eltwise.add"(%0, %cst) : (tensor<3x3x!eltwise.f64>, !f32) -> tensor<3x3x!eltwise.f64>
+    return %1 : tensor<3x3x!eltwise.f64>
```

Comment:
Again, I would expect this to return `f32`

Reply:
done

Reply:
Isn't this still returning an `f64`? Shouldn't it return an `f32`?

Reply:
Sorry about the incorrect previous comment. Maybe I misunderstood our previous conversation, but here, the user passes `float64` as a Placeholder, and conflicting type `floatx=float32.` Type promotion yields return `max(float32, float64) => float64`

Reply:
I guess I'm confused about why we'd want to have tests for situations where `floatx` and the input datatypes don't match. I think in that situation the user is requesting conflicting types but we should always choose the most promoted type between the two to resolve the conflict. However, the `floatx` API and the new `ConstantTypesPass` should allow the user to specify some other datatype (other than the default=`float32`) such that the result type is ensured to be that other datatype and that no truncation would occur.